lib/improver/utilities/warnings_handler.py

```diff
@@ -0,0 +1,132 @@
+# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+"""Classes and functions for managing warnings in the IMPROVER code."""
+
+import warnings
+import sys
+
+
+class ManageWarnings(object):
+    """
+    A decorator used to manage the warnings that are raised by a function.
+    Ignore a selection of warnings, and either raise any remaining warnings
+    to standard error or record them in a list of warning objects.
+    """
+    def __init__(self, ignored_messages=None, warning_types=None,
+                 record=False):
+        """
+        Set up a decorator with the warnings we want to ignore and what
+        we want to do with any remaining warnings.
+
+        Keyword Args:
+            ignored_messages (List of strings):
+                A list of messages, one for each warning message we
+                want to ignore.
+            warning_types (List):
+                A list containing the Warning category for each of the
+                messages. If not provided then the Warning Category is
+                assumed to be UserWarning for each of the messages.
+            record (Boolean):
+                A flag for whether to store any warnings that are not
+                ignored. Default is False which means warnings go to
+                standard error. When set to True the warnings are
+                recorded in a warning list which is passed to the function
+                being decorated.
+
+        Raises:
+            TypeError: ignored_messages not list.
+            ValueError: Raise error if both ignored messages and warning_types
+                        are given and they are not the same length.
+        """
+        if ignored_messages is not None:
+            if not isinstance(ignored_messages, list):
+                msg = 'Expecting list of strings for ignored_messages'
+                raise TypeError(msg)
+        self.messages = ignored_messages
+        if warning_types is None and self.messages is not None:
+            self.warning_types = [UserWarning for _ in self.messages]
+        else:
+            self.warning_types = warning_types
+        self.record = record
+        if self.messages and (len(self.warning_types) != len(self.messages)):
+            message = ("Length of warning_types ({}) does no equal length"
+                       "of warning messages({})")
+            message = message.format(len(self.warning_types),
+                                     len(self.messages))
+            raise ValueError(message)
+
+    def reset_warning_registry(self):
```

Comment:
This doesn't have any unit tests. Do you think we will be able to write one or will we need to leave it because we can't write a suitable test?

Reply:
No, I  would't have the first idea how to unit test this